### PR TITLE
Fix fetch resume

### PR DESCRIPTION
**Pull request by piotro888**
Fixes bug found in #654
Splits `fetch.resume` to separate `resume_from_exception` and `resume_from_unsafe`

I tried adding formal-like verification constraints, to verify method ordering assumptions from the comments, but have mixed feelings on the final solution. What do you think?

### CODE REVIEW

File: coreblocks/core.py [hunk 1]
```diff
@@ -195,4 +194,7 @@ def elaborate(self, platform):
             read = self.connections.get_dependency(BranchVerifyKey())
             read(m)  # Consume to not block JB Unit
 
+        if self.gen_params.extra_verification:
+            m.d.comb += self.fetch._verification_resume_from_collector_ready.eq(fb_fetch_resume.ready)
```
Comment (tilk) [inline coreblocks/core.py:198]: Can this be done in `FetchUnit` instead?

Comment (tilk) [reply coreblocks/core.py:198]: This is related to `UnfiierKey` handling in `DependencyManager`, left for future PR.

File: coreblocks/frontend/fetch/fetch.py [hunk 2]
```diff
@@ -367,21 +367,80 @@ def flush():
...
+                    m.d.sync += expect_unstall_unsafe.eq(1)
+                    with m.If(self.resume_from_unsafe.run):
+                        m.d.sync += expect_unstall_unsafe.eq(0)
+                        m.d.sync += prev_stalled_unsafe.eq(0)  # it is fine to be stalled now
+                        m.next = "running"
+                    with m.If(self.stall_exception.run):
+                        m.next = "stalled_exception"
+                    log.error(
+                        m,
+                        self.resume_from_exception.run & ~self.stall_exception.run,
```
Comment (lekcyjna123) [inline coreblocks/frontend/fetch/fetch.py:429]: `~self.stall_exception.run` is always `False` here

File: coreblocks/frontend/fetch/fetch.py [hunk 3]
```diff
@@ -367,21 +367,80 @@ def flush():
...
+                        "unexpected resume_from_exception",
+                    )
+                with m.State("stalled_exception"):
+                    with m.If(self.resume_from_unsafe.run):
+                        log.error(m, ~expect_unstall_unsafe, "unexpected resume_from_unsafe")
+                        m.d.sync += expect_unstall_unsafe.eq(0)
+                    with m.If(self.resume_from_exception.run):
+                        # unstall_form_unsafe may be skipped if excpetion was reported on unsafe instruction,
+                        # invlid cases are verified by readiness check in running state
```
Comment (lekcyjna123) [inline coreblocks/frontend/fetch/fetch.py:438]:
```suggestion
                        # invalid cases are verified by readiness check in running state
```

**Comment by xThaid** [comment]
I agree that having verification code is a good idea. Perhaps we could make another module (something like `FetchVerification`) that contains that logic to avoid polluting the main module, but then we need to expose some signals, so it is not perfect. Another option would be to create a function within the main module that is called from `elaborate` and contains the verification logic.

File: coreblocks/core.py (hunk 1 above)
Comment (piotro888) [reply coreblocks/core.py]: fixed with #688

File: coreblocks/frontend/fetch/fetch.py (hunk 2 above)
Comment (piotro888) [reply coreblocks/frontend/fetch/fetch.py:420]: why? it is not under the If

**Review by tilk — APPROVED** [review]

**Review by lekcyjna123 — APPROVED** [review]

**Pull request merged**